llvm/lib/Transforms/Utils/ControlFlowUtils.cpp

```diff
@@ -0,0 +1,341 @@
+      }
+      NewPhi->addIncoming(V, BB);
+    }
+    assert(NewPhi->getNumIncomingValues() == Incoming.size());
+    Value *NewV = NewPhi;
+    if (AllUndef) {
+      NewPhi->eraseFromParent();
+      NewV = UndefValue::get(Phi->getType());
+    }
+    if (Phi->getNumOperands() == 0) {
+      Phi->replaceAllUsesWith(NewV);
+      I = Phi->eraseFromParent();
+      continue;
+    }
+    Phi->addIncoming(NewV, GuardBlock);
+    ++I;
+  }
+}
+
+BasicBlock *ControlFlowHub::finalize(
+    DomTreeUpdater *DTU, SmallVectorImpl<BasicBlock *> &GuardBlocks,
+    const StringRef Prefix, std::optional<unsigned> MaxControlFlowBooleans) {
+#ifndef NDEBUG
+  SmallSet<BasicBlock *, 8> Incoming;
+#endif
+  SetVector<BasicBlock *> Outgoing;
+
+  for (auto [BB, Succ0, Succ1] : Branches) {
+#ifndef NDEBUG
+    assert(Incoming.insert(BB).second && "Duplicate entry for incoming block.");
+#endif
+    if (Succ0)
+      Outgoing.insert(Succ0);
+    if (Succ1)
+      Outgoing.insert(Succ1);
+  }
+
+  if (Outgoing.size() < 2)
+    return Outgoing.front();
+
+  SmallVector<DominatorTree::UpdateType, 16> Updates;
+  if (DTU) {
+    for (auto [BB, Succ0, Succ1] : Branches) {
+      if (Succ0)
+        Updates.push_back({DominatorTree::Delete, BB, Succ0});
+      if (Succ1)
+        Updates.push_back({DominatorTree::Delete, BB, Succ1});
+    }
+  }
+
+  SmallVector<WeakVH, 8> DeletionCandidates;
+  convertToGuardPredicates(Branches, Outgoing.getArrayRef(), GuardBlocks,
+                           DeletionCandidates, Prefix, MaxControlFlowBooleans);
+  BasicBlock *FirstGuardBlock = GuardBlocks.front();
+
+  // Update the PHINodes in each outgoing block to match the new control flow.
+  for (int I = 0, E = GuardBlocks.size(); I != E; ++I)
+    reconnectPhis(Outgoing[I], GuardBlocks[I], Branches, FirstGuardBlock);
+  // Process the Nth (last) outgoing block with the (N-1)th (last) guard block.
+  reconnectPhis(Outgoing.back(), GuardBlocks.back(), Branches, FirstGuardBlock);
+
+  if (DTU) {
+    int NumGuards = GuardBlocks.size();
+
+    for (auto [BB, Succ0, Succ1] : Branches)
+      Updates.push_back({DominatorTree::Insert, BB, FirstGuardBlock});
+
+    for (int I = 0; I != NumGuards - 1; ++I) {
+      Updates.push_back({DominatorTree::Insert, GuardBlocks[I], Outgoing[I]});
+      Updates.push_back(
+          {DominatorTree::Insert, GuardBlocks[I], GuardBlocks[I + 1]});
+    }
+    // The second successor of the last guard block is an outgoing block instead
+    // of having a "next" guard block.
+    Updates.push_back({DominatorTree::Insert, GuardBlocks[NumGuards - 1],
+                       Outgoing[NumGuards - 1]});
+    Updates.push_back({DominatorTree::Insert, GuardBlocks[NumGuards - 1],
+                       Outgoing[NumGuards]});
+    DTU->applyUpdates(Updates);
+  }
+
+  for (auto I : DeletionCandidates) {
+    if (I->use_empty())
+      if (auto *Inst = dyn_cast_or_null<Instruction>(I))
+        Inst->eraseFromParent();
+  }
```

Comment:
I guess this isnt' a change, but does this really need to bother cleaning this up?

Reply:
It's nice to cleanup while we can, and results in cleaner lit tests since the unused PHIs show up as removed in the checks.